README.md
# sscg

Generate self signed certificate

## Usage

```sh
sscg - 0.0.0

Generate self signed certificate.

  -ca
   	whether this cert should be its own Certificate Authority
  -duration duration
   	Duration that certificate is valid for (default 8760h0m0s)
  -ecdsa-curve string
   	ECDSA curve to use to generate a key. Valid values are P224, P256, P384, P521
  -h
  -help
   	Get help
  -host string
   	Comma-separated hostnames and IPs to generate a certificate for
  -out string
   	Output put directory (default ".")
  -rsa-bits int
   	Size of RSA key to generate. Ignored if --ecdsa-curve is set (default 2048)
  -start-date string
   	Creation date formatted as Jan 1 15:04:05 2011
  -v
  -version
   	Get version
```

__Examples__

```sh
  sscg --host dev.example.org
  sscg --help
  sscg --version
```

## Install

Pick an msi package [here](https://github.com/mh-cbon/sscg/releases)!

__chocolatey__

```sh
choco install sscg
```

__deb/rpm repositories__

```sh
wget -O - https://raw.githubusercontent.com/mh-cbon/latest/master/source.sh \
| GH=mh-cbon/sscg sh -xe
# or
curl -L https://raw.githubusercontent.com/mh-cbon/latest/master/source.sh \
| GH=mh-cbon/sscg sh -xe
```

__deb/rpm packages__

```sh
curl -L https://raw.githubusercontent.com/mh-cbon/latest/master/install.sh \
| GH=mh-cbon/sscg sh -xe
# or
wget -q -O - --no-check-certificate \
https://raw.githubusercontent.com/mh-cbon/latest/master/install.sh \
| GH=mh-cbon/sscg sh -xe
```

__go__

```sh
mkdir -p $GOPATH/src/github.com/mh-cbon
cd $GOPATH/src/github.com/mh-cbon
git clone https://github.com/mh-cbon/sscg.git
cd sscg
glide install
go install
```

## Credits

Orignal code taken from https://golang.org/src/crypto/tls/generate_cert.go and re-packed.


CHANGELOG.md
# Changelog - sscg

### 0.0.4

__Changes__

- deply: fix GH_TOKEN for travis

__Contributors__

- mh-cbon

Released by mh-cbon, Thu 03 Nov 2016 -
[see the diff](https://github.com/mh-cbon/sscg/compare/0.0.3...0.0.4#diff)
______________

### 0.0.3

__Changes__

- deploy: fix travis release key

__Contributors__

- mh-cbon

Released by mh-cbon, Thu 03 Nov 2016 -
[see the diff](https://github.com/mh-cbon/sscg/compare/0.0.2...0.0.3#diff)
______________

### 0.0.2

__Changes__

- deploy: fix typos in both appveyor / travis files

__Contributors__

- mh-cbon

Released by mh-cbon, Thu 03 Nov 2016 -
[see the diff](https://github.com/mh-cbon/sscg/compare/0.0.1...0.0.2#diff)
______________

### 0.0.1

__Changes__

- init

__Contributors__

- mh-cbon

Released by mh-cbon, Thu 03 Nov 2016 -
[see the diff](https://github.com/mh-cbon/sscg/compare/cb711a8af9bc509445bd8330519900534835b17f...0.0.1#diff)
______________




test.sh
#!/bin/sh
rm -fr test/

go run main.go --host example.org --out test

(openssl x509 -in test/cert.pem -text -noout | grep "DNS:example.org" && echo "OK") || exit 1
(openssl x509 -in test/cert.pem -text -noout | grep "Issuer: O=Acme Co" && echo "OK") || exit 1
(openssl x509 -in test/cert.pem -text -noout | grep "Subject: O=Acme Co" && echo "OK") || exit 1

echo "test passed!"


main.go
// Copyright 2009 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// Generate a self-signed X.509 certificate for a TLS server. Outputs to
// 'cert.pem' and 'key.pem' and will overwrite existing files.

package main

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/rsa"
	"crypto/x509"
	"crypto/x509/pkix"
	"encoding/pem"
	"flag"
	"fmt"
	"log"
	"math/big"
	"net"
	"os"
	"path"
	"strings"
	"time"
)

var (
	VERSION    = "0.0.0"
	v          = flag.Bool("version", false, "Get version")
	ver        = flag.Bool("v", false, "")
	help       = flag.Bool("help", false, "Get help")
	h          = flag.Bool("h", false, "")
	out        = flag.String("out", ".", "Output put directory")
	host       = flag.String("host", "", "Comma-separated hostnames and IPs to generate a certificate for")
	validFrom  = flag.String("start-date", "", "Creation date formatted as Jan 1 15:04:05 2011")
	validFor   = flag.Duration("duration", 365*24*time.Hour, "Duration that certificate is valid for")
	isCA       = flag.Bool("ca", false, "whether this cert should be its own Certificate Authority")
	rsaBits    = flag.Int("rsa-bits", 2048, "Size of RSA key to generate. Ignored if --ecdsa-curve is set")
	ecdsaCurve = flag.String("ecdsa-curve", "", "ECDSA curve to use to generate a key. Valid values are P224, P256, P384, P521")
)

func publicKey(priv interface{}) interface{} {
	switch k := priv.(type) {
	case *rsa.PrivateKey:
		return &k.PublicKey
	case *ecdsa.PrivateKey:
		return &k.PublicKey
	default:
		return nil
	}
}

func pemBlockForKey(priv interface{}) *pem.Block {
	switch k := priv.(type) {
	case *rsa.PrivateKey:
		return &pem.Block{Type: "RSA PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(k)}
	case *ecdsa.PrivateKey:
		b, err := x509.MarshalECPrivateKey(k)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Unable to marshal ECDSA private key: %v", err)
			os.Exit(2)
		}
		return &pem.Block{Type: "EC PRIVATE KEY", Bytes: b}
	default:
		return nil
	}
}

func main() {
	flag.Parse()

	if *h || *help {
		fmt.Printf("sscg - %v\n", VERSION)
		fmt.Println("")
		fmt.Println("Generate self signed certificate.")
		fmt.Println("")
		flag.PrintDefaults()
		fmt.Println("")
		return
	}

	if *ver {
		fmt.Printf("%v\n", VERSION)
		return
	}

	if len(*host) == 0 {
		log.Fatalf("Missing required --host parameter")
	}

	var outDirectory = ""
	if len(*out) > 0 {
		outDirectory = *out
	}

	var priv interface{}
	var err error
	switch *ecdsaCurve {
	case "":
		priv, err = rsa.GenerateKey(rand.Reader, *rsaBits)
	case "P224":
		priv, err = ecdsa.GenerateKey(elliptic.P224(), rand.Reader)
	case "P256":
		priv, err = ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	case "P384":
		priv, err = ecdsa.GenerateKey(elliptic.P384(), rand.Reader)
	case "P521":
		priv, err = ecdsa.GenerateKey(elliptic.P521(), rand.Reader)
	default:
		fmt.Fprintf(os.Stderr, "Unrecognized elliptic curve: %q", *ecdsaCurve)
		os.Exit(1)
	}
	if err != nil {
		log.Fatalf("failed to generate private key: %s", err)
	}

	var notBefore time.Time
	if len(*validFrom) == 0 {
		notBefore = time.Now()
	} else {
		notBefore, err = time.Parse("Jan 2 15:04:05 2006", *validFrom)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Failed to parse creation date: %s\n", err)
			os.Exit(1)
		}
	}

	notAfter := notBefore.Add(*validFor)

	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
	if err != nil {
		log.Fatalf("failed to generate serial number: %s", err)
	}

	template := x509.Certificate{
		SerialNumber: serialNumber,
		Subject: pkix.Name{
			Organization: []string{"Acme Co"},
		},
		NotBefore: notBefore,
		NotAfter:  notAfter,

		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
		BasicConstraintsValid: true,
	}

	hosts := strings.Split(*host, ",")
	for _, h := range hosts {
		if ip := net.ParseIP(h); ip != nil {
			template.IPAddresses = append(template.IPAddresses, ip)
		} else {
			template.DNSNames = append(template.DNSNames, h)
		}
	}

	if *isCA {
		template.IsCA = true
		template.KeyUsage |= x509.KeyUsageCertSign
	}

	derBytes, err := x509.CreateCertificate(rand.Reader, &template, &template, publicKey(priv), priv)
	if err != nil {
		log.Fatalf("Failed to create certificate: %s", err)
	}

	if _, err := os.Stat(outDirectory); os.IsNotExist(err) {
		os.Mkdir(outDirectory, 0700)
	}

	certPath := path.Join(outDirectory, "cert.pem")
	certOut, err := os.Create(certPath)
	if err != nil {
		log.Fatalf("failed to open '%s' for writing: %s", certPath, err)
	}
	pem.Encode(certOut, &pem.Block{Type: "CERTIFICATE", Bytes: derBytes})
	certOut.Close()
	log.Printf("written '%s'\n", certPath)

	keyPath := path.Join(outDirectory, "key.pem")
	keyOut, err := os.OpenFile(keyPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
	if err != nil {
		log.Fatalf("failed to open '%s' for writing: %s\n", keyPath, err)
	}
	pem.Encode(keyOut, pemBlockForKey(priv))
	keyOut.Close()
	log.Printf("written '%s'\n", keyPath)
}
